### ISSUE TITLE
Schema Validation: it's ok to be null

### ISSUE THREAD

**Issue by bleonard** [airbytehq/airbyte#24987]
A huge majority of the `oneOf` / `type array` used in connector schemas is to say that the value could also be `null`. These tend to look like [this](https://github.com/airbytehq/airbyte/blob/ceadf31444dcd1b6e15796c29e1cd6928cbefb97/airbyte-integrations/connectors/source-bigcommerce/integration_tests/catalog.json):

```
    "postal_code": {
      "type": ["null", "string"]
    },
    "accepts_product_review_abandoned_cart_emails": {
      "type": ["null", "boolean"]
    },
    "store_credit_amounts": {
      "type": ["null", "array"],
      "items": {
        "type": ["null", "object"],
        "properties": {
          "amount": {
            "type": ["null", "number"]
          }
        }
      }
    },
```

Other places use `oneOf` like [this](https://github.com/airbytehq/airbyte/blob/09955423dba22586131364485d8047b429ba6db9/airbyte-integrations/connectors/source-flexport/source_flexport/schemas/shared/shipment.json):

```
    "ocean_shipment": {
  "description": "...",
  "oneOf": [
    {
      "type": "null"
    },
    {
      "$ref": "ocean/shipment.json"
    }
  ]
},
```

We have had [bugs](https://github.com/airbytehq/airbyte/issues?q=is%3Aissue+%22null+found%22++expected+%22JSON+schema+validation%22) where we forget to add the `null` to the schema, leave the key/value pair in (as `null`) and JSONSchema validation fails. On the other hand, sources can also just not emit the key/value and things are fine.

We should simplify the world and clarify the current reality: that any value can be `null`.

This would allow the above schema to look like this:

```
    "postal_code": {
      "type": "string"
    },
    "accepts_product_review_abandoned_cart_emails": {
      "type": "boolean"
    },
    "store_credit_amounts": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "amount": {
            "type": "number"
          }
        }
      }
    },
```

And for the other one:

```
    "ocean_shipment": {
      "description": "...",
      "$ref": "ocean/shipment.json"
    },
```

Overall, this allows less nuances in schemas and more confident connector building.

Also, by removing our most common usage of the `type array`, we can move one step closer to standardizing between that and `oneOf`.

## Open Questions

1. Should there be a way to opt into validating for null? Maybe primary or foreign keys?

    ```
    "postal_code": {
      "type": "string"
      "nullable": "false"
    },

**Comment by cgardens** [comment]
@bleonard could we get best of both world by saying don't put `null` in the JSONSchema `type` field, but also don't emit null (i.e. just don't include the field at all)?

**Comment by evantahler** [comment]
I'm going to close this issue in favor of the per-source work that @prateekmukhedkar is going to do.  @prateekmukhedkar can you link an epic for the sources to support NULLs?

**Comment by bleonard** [comment]
What's the plan here? I don't think 1000 sources should have to do `type: [null, string]` for every field in order to make sure we have primary keys. There has to be another way forward.

Allowing null is the current behavior the platform has because it's not verifying fields that are left off by the source. Are we going to start failing those records?

**Comment by evantahler** [comment]
> Are we going to start failing those records?
I don't think there's a short-term plan for this, but, yes - one day I think we should!

**Comment by cgardens** [comment]
@bleonard I have a PR up to do this in the Platform. I paused it because Ella was in-progress on another validation project that touched the same part of the code base. Once she is done, I can rebase the PR and merge it.

That said, sounds like there's some disagreement on how approach here. Happy to adapt.

**Comment by evantahler** [comment]
@bleonard and I chatted, and closing this issue was what we landed on. `[null]` means something, and for now, we will go forward down the road of changing connectors to include `[null]` if they really mean it.  This information is too useful to stub out, as PKs, FKs, and cursors should all probably be non-null, and we want this information for the platform and destination for future checks and error handling. 

In the future, when we add new datatypes, we can consider adding a new record property option, like `isNullable:bool` to make this more ergonomic.
